Handle simple get requests

MaggieVL opened the pull request.

Commit:
Add .gitignore

Commit:
Add elmForAPI function with generator parameter

Commit:
Resolve readme and license conflicts

Commit:
Remove generator parameter

Commit:
Add dependencies

elm-bridge is needed temporarily until elm-street is used thouroughly
instead of it

Commit:
Add imports and extensions

Commit:
Add function generating query from an endpoint

This is a first still not working version. Changes need to be done.
Elm-bridge needs to be replaced with elm-street

Commit:
Fix type signature

The url prefix we need is only dynamic so I removed the static part.

Commit:
Add Servant.Elm functions

This is a version of the project that is not working but is full for
generating a query, using mainly Servant.Elm functions.

Commit:
Modify .gitignore

Commit:
Remove unused code

This is all the code from Servant.Elm that I did not need. There are
still some patches from other places that couldn't be omitted.

Commit:
Create the url

Commit:
Add function to create function name

Commit:
Implement elmTypeRef

Commit:
Add indent function

Commit:
Extract method from Req ElmDefinition

Commit:
Add main

Commit:
Add imports

googleson78 commented on .gitignore:
```diff
@@ -21,3 +21,5 @@ cabal.project.local
 .HTF/
 .ghc.environment.*
+*~
+src/copy/
```
What's this for?

MaggieVL replied:
there was a copy of the project because I hadn't committed a few important changes but I don't need now

googleson78 replied:
So you should remove this from the `.gitignore`

googleson78 commented on ChangeLog.md:
```diff
@@ -0,0 +1,3 @@
+# Changelog for servant-elm-street
+
+## Unreleased changes
```
We don't need this file for now, I think you should delete it.

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+name:                servant-elm-street
+version:             0.1.0.0
+github:              "githubuser/servant-elm-street"
```
Fix this

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+version:             0.1.0.0
+github:              "githubuser/servant-elm-street"
+license:             BSD3
+author:              "Author name here"
+maintainer:          "example@example.com"
+copyright:           "2021 Author name here"
```
Fix these. Not sure what they should be though.

googleson78 replied:
@dtaskoff please verify this when you're here

dtaskoff replied:
The copyright should state "Copyright 2022 Chaos Software OOD". The others fields can be filled as you like.

MaggieVL replied:
done

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+
+extra-source-files:
+- README.md
+- ChangeLog.md
```
If we remove the changelog file, this line should also go.

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+- README.md
+- ChangeLog.md
+
+# Metadata used when publishing your package
+# synopsis:            Short description of your package
+# category:            Web
+
+# To avoid duplicated efforts in documentation and dealing with the
+# complications of embedding Haddock markup inside cabal files, it is
+# common to point users to the README.md file.
```
Should delete this generated code.

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+# To avoid duplicated efforts in documentation and dealing with the
+# complications of embedding Haddock markup inside cabal files, it is
+# common to point users to the README.md file.
+description:         Please see the README on GitHub at <https://github.com/githubuser/servant-elm-street#readme>
```
Fix this

googleson78 commented on stack.yaml:
```diff
@@ -0,0 +1,70 @@
+# This file was automatically generated by 'stack init'
+#
+# Some commonly used options have been documented as comments in this file.
+# For advanced use and comprehensive documentation of the format, please see:
+# https://docs.haskellstack.org/en/stable/yaml_configuration/
+
+# Resolver to choose a 'specific' stackage snapshot or a compiler version.
+# A snapshot resolver dictates the compiler version and the set of packages
+# to be used for project dependencies. For example:
+#
+# resolver: lts-3.5
+# resolver: nightly-2015-09-21
+# resolver: ghc-7.10.2
+#
+# The location of a snapshot can be provided as a file or url. Stack assumes
+# a snapshot provided as a file might change, whereas a url resource does not.
+#
+# resolver: ./custom-snapshot.yaml
+# resolver: https://example.com/snapshots/2018-01-01.yaml
```
We don't need all the generated comments in this file.

googleson78 commented on app/Main.hs:
```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE DataKinds #-}
```
Since you're only using this for testing, I don't think we should be committing this file.

googleson78 replied:
As we discussed, it should instead be moved to an `examples` folder in a project of its own.

MaggieVL replied:
done

googleson78 commented on Setup.hs:
```diff
@@ -0,0 +1,2 @@
+import Distribution.Simple
```
This file should be safe to delete

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+  source-dirs: src
+
+dependencies:
+- servant-foreign
+- text
+- servant-elm
+- base >= 4.7 && < 5
+- aeson
+- casing
+- elm-street
+- filepath
+- servant
+- servant-server
+- wai
+- warp
+- servant-js
+- random
+- elm-bridge
+- wl-pprint-text
+- microlens
```
A lot of these dependencies are only used by your example API that you're using to run your code, once you remove it, you should also delete them.

googleson78 replied:
You can at least move all the things that are only used by the example API to the `dependencies` section of the executable, since they're only used there.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,215 @@
+          dquotes (stext (unPathSegment path))
+        Cap arg ->
+          error
+            "TODO implement-2" -- for captures, not needed now
```
Just put the comment inside the error, no point in having some other text in the error which isn't very informative.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,215 @@
+  RefPrim elmPrim -> elmPrimToDoc elmPrim
+  RefCustom (TypeName typeName) -> pretty typeName
+
+elmTypeParenDoc :: TypeRef -> Doc -- does it make sense to be a fn
```
Why not? It's used more than once, so it seems fine to have this.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,215 @@
+
+mkTypeSignature :: Req ElmDefinition -> Doc
+mkTypeSignature request =
+  (hsep . punctuate " ->") ("String" : catMaybes [toMsgType, returnType])
```
You should usually not be manually inserting spaces in strings when using a pretty printer, like in `" ->"` - shouldn't `hsep` already be taking care to put the spaces anyway?

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,215 @@
+    elmTypeRef eDef = elmTypeRefToDoc (definitionToRef eDef)
+    toMsgType :: Maybe Doc
+    toMsgType = do
+      result <- fmap elmTypeRef $ request ^. reqReturnType
+      Just ("(Result Http.Error " <+> parens result <+> " -> msg)")
```
You don't need the `do` block and the bind here, you can do it with a single `fmap`, I think
```haskell
toMsgType = 
  fmap mkMsgType $ request ^. reqReturnType
  where
    mkMsgType x = ("(Result Http.Error " <+> parens (elmTypeRef x) <+> " -> msg)")
```

Also, try not to manually add spaces, like in `"-> msg)"`

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,215 @@
+      Just ("(Result Http.Error " <+> parens result <+> " -> msg)")
+
+    returnType :: Maybe Doc
+    returnType = do
```
Redundant `do`

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,215 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -Wno-deferred-out-of-scope-variables #-}
```
Highly doubtful we need this.

googleson78 replied:
Try removing it.

MaggieVL replied:
done

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,70 @@
+- elm-bridge
+- wl-pprint-text
+- microlens
+
```
We should turn on more warnings and error out on warnings. Here's a snippet to put in this file:
```yaml
ghc-options:
- -Wall
- -Werror
- -Wincomplete-uni-patterns
- -Wredundant-constraints
- -Widentities
- -Wincomplete-record-updates
- -Wmissing-deriving-strategies
- -Wmissing-export-lists
```

googleson78 replied:
(of course you should also recompile afterwards to fix all the warnings)

MaggieVL replied:
done

Commit:
Add simple expect and renderDecoderName

This is a version that takes ElmDefinition, but we need it to take a
TypeRef because we can't use recursion on all types now.

Commit:
Delete files that are not needed

Commit:
Modify .gitignore

Commit:
Modify package.yaml

Change description, dependencies and add warnings

Commit:
Modify servant-elm-street.cabal

Commit:
Remove comments

Commit:
Remove unnecessary do block

Commit:
Remove redundant do and mkArgs

Commit:
Remove redundant space

Commit:
Fix comments in mkUrl

googleson78 commented on package.yaml:
```diff
@@ -1,48 +1,41 @@
 - wai
-- warp
-- servant-js
 - random
```
What do you need `random` f or?

MaggieVL replied:
it was for an experiment long ago, I removed it now

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -172,7 +172,7 @@ elmPrimToDoc = \case
 mkTypeSignature :: Req ElmDefinition -> Doc
 mkTypeSignature request =
-  (hsep . punctuate " ->") ("String" : catMaybes [toMsgType, returnType])
+  (hsep . punctuate "->") ("String" : catMaybes [toMsgType, returnType])
```
Did you verify that this doesn't cause an issue?

MaggieVL replied:
it does cause an issue, I added the space back

MaggieVL replied:
done

googleson78 replied:
Let's leave it like this for now, because I don't think it will be an issue, but in general, the "correct" way to do this would be something like
```haskell
let msgs :: [Doc ann]
let arr = "->"
hsep $ intersperse arr msgs
```
I think.

googleson78 replied:
Also, in general, when making a change, if it isn't too difficult, you should test out if your change breaks anything, before committing it (and try to fix things, if it does)

Commit:
Fix import warnings

Commit:
Use the Prettyprinter used in elm-street

Commit:
Move Main to examples directory

All the example code I use to test ServantElm needs to be in a separate
place because it's not part of the project.

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,231 @@
+    vsep,
+    (<+>),
+  )
+import Prettyprinter.Internal (unsafeTextWithoutNewlines)
```
Do you need this? From the docs, it has an precondition that must be met. Do you know if it's met where you're using it?   I'm pretty sure there's a non-"unsafe" way to do this.

googleson78 replied:
e.g. you can use the `Pretty` class for this, I think. There's an instance for `Text`.

https://hackage.haskell.org/package/prettyprinter-1.7.1/docs/Prettyprinter.html#t:Pretty

MaggieVL replied:
done

googleson78 commented on examples/servant-elm-street-main/package.yaml:
```diff
@@ -0,0 +1,42 @@
+    dependencies:
+    - servant-elm-street-main
+
+tests:
```
The test suite is unused, you can delete this.

MaggieVL replied:
done

googleson78 commented on examples/servant-elm-street-main/LICENSE:
```diff
@@ -0,0 +1,30 @@
+Copyright Author name here (c) 2022
```
This file isn't necessary imo

MaggieVL replied:
done

googleson78 commented on examples/servant-elm-street-main/.gitignore:
```diff
@@ -0,0 +1 @@
+.stack-work/
```
This file gets autogenerated by stack, you can delet it from the example

MaggieVL replied:
done

googleson78 commented on examples/servant-elm-street-main/stack.yaml:
```diff
@@ -0,0 +1,9 @@
+- .
+
+extra-deps:
+- /home/magdalena/projects/servant-playground/servant-elm-street
```
Use a relative path so that
* this works for others
* it doesn't contain information about your computer

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,231 @@
+    reqReturnType,
+    reqUrl,
+  )
+import Prelude hiding ((<$>))
```
You no longer need to hide this now that you've switched over to the other pretty printer.

MaggieVL replied:
done

Commit:
Change author fields

Commit:
Remove unused dependencies

Commit:
Move main in examples directory

Commit:
Compile main

googleson78 commented on .gitignore:
```diff
@@ -21,3 +21,4 @@ cabal.project.local
 cabal.project.local~
 .HTF/
 .ghc.environment.*
+personal-access-token-git.txt
```
What are you using this for?

Commit:
Change copyright year

Commit:
Use safe version for Text to Doc fn

Commit:
Remove test suite

Commit:
Stop hiding <$> from Prelude

We don't use it any more so we don't need to hide it.

Commit:
Delete unused test suite

Commit:
Fix missing whitespaces in fn definitions

Commit:
Remove unused files

Commit:
Remove unused dependency

Commit:
Remove license from .cabal

Commit:
Switch from expectJson to expectWhatever

We will handle the response body in the future so we don't need it now.

googleson78 commented on package.yaml:
```diff
@@ -0,0 +1,45 @@
+- -Wmissing-export-lists
+- -Wunused-packages
+
+tests:
```
Unused, should delete this.

googleson78 commented on examples/servant-elm-street-main/test/Spec.hs:
```diff
@@ -0,0 +1,2 @@
+main :: IO ()
```
Unused file, delete this.

googleson78 commented on test/Spec.hs:
```diff
@@ -0,0 +1,2 @@
+main :: IO ()
```
Unused file, delete this.

Commit:
Fix dependencies

Commit:
Remove test suites

googleson78 commented on src/ServantElm.hs:
```diff
@@ -0,0 +1,171 @@
+      (map segmentToDoc segments)
+  where
+    urlBuilder :: Doc ann
+    urlBuilder = "Url.Builder.crossOrigin urlBase" :: Doc ann
```
This annotation here should not be needed, because you've already given a type signature above.

MaggieVL replied:
done

Commit:
Remove unnecessary type specification
